src/main/scala/com/fulcrumgenomics/bam/DownsampleAndNormalizeBam.scala

```diff
@@ -0,0 +1,206 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2022 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.bam.DownsampleAndNormalizeBam.CoverageTracker
+import com.fulcrumgenomics.bam.api._
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.commons.util.LazyLogging
+import com.fulcrumgenomics.sopt.{arg, clp}
+import com.fulcrumgenomics.util.{Io, ProgressLogger, Sorter}
+import htsjdk.samtools.util.IntervalList.IntervalMergerIterator
+import htsjdk.samtools.util.{CoordMath, Interval, IntervalList, Murmur3, OverlapDetector}
+
+/**
+  * Companion object with helper classes.
+  */
+private object DownsampleAndNormalizeBam extends LazyLogging {
+  /**
+    * Tracks coverage over a set of regions.
+    */
+  class CoverageTracker(regions: Seq[Interval],
+                        val targetCoverage: Int,
+                        val minMapQ: Int
+                       ) {
+
+    // Overlap detector for region => coverage
+    private val detector = new OverlapDetector[RegionCoverage](0, 0)
+    regions.foreach { region =>
+      detector.addLhs(new RegionCoverage(region.getContig, region.getStart, region.getEnd), region)
+    }
+
+    /**
+      * Checks to see if a template has reads that contribute coverage to any bases that are currently
+      * below the coverage threshold.  If so, coverage from all acceptable reads is accumulated and the
+      * method returns true.  Otherwise no coverage is accumulated and the method returns false.
+      *
+      * @param t a Template to accumulate coverage from if it covers any bases that are under the coverage threshold
+      * @return true if the template added coverage, false otherwise
+      */
+    def add(t: Template): Boolean = {
+      val recs = t.allReads
+        .filterNot(_.secondary)
```

Comment:
This made me think about supplementary reads.  Should those be upgraded to "primary" alignments or be filtered out as well?  I'm thinking of a region where we have a large number of both primary and supplementary alignments, and how to pick there.

Reply:
I definitely want to include supplementary reads, they're just part of a split-read alignment and therefore contribute to coverage.  E.g. imagine an alignment to a circular genome where anything that maps over the break point will have a primary+supplementary, we want to count both of those.